src/java/org/apache/sqoop/avro/AvroUtil.java

```diff
@@ -114,11 +114,20 @@ public static String toAvroColumn(String column) {
    * Format candidate to avro specifics
    */
   public static String toAvroIdentifier(String candidate) {
-    String formattedCandidate = candidate.replaceAll("\\W+", "_");
-    if (formattedCandidate.substring(0,1).matches("[a-zA-Z_]")) {
-      return formattedCandidate;
+    char[] data = candidate.toCharArray();
+    int stringIndex = 0;
+
+    for (char c:data) {
+      if (Character.isLetterOrDigit(c) || c == '_') {
+        data[stringIndex++] = c;
+      }
+    }
+
+    char initial = data[0];
+    if (Character.isLetter(initial) || initial == '_') {
+      return new String(data, 0, stringIndex);
```

Comment:
Your code will first create a char array and then eventually update char in the array. As result you will create another copy as a new String. Have you thought about using a `StringBuilder` directly?
```
  final StringBuilder sb = new StringBuilder();
  for (char c : candidate) {
    if (Character.isLetterOrDigit(c) || c == '_') {
      sb.append(c);
    }
  }
  ...
    return sb.toString();
```